src/lib.rs

```diff
@@ -2528,34 +2528,63 @@ pub trait Itertools: Iterator {
     ///
     /// If `f` is associative you should also decide carefully:
     ///
-    /// - if `f` is a trivial operation like `u32::wrapping_add`, prefer the normal
-    /// [`Iterator::reduce`] instead since it will most likely result in the generation of simpler
-    /// code because the compiler is able to optimize it
-    /// - otherwise if `f` is non-trivial like `format!`, you should use `tree_reduce` since it
-    /// reduces the number of operations from `O(n)` to `O(ln(n))`
```

Comment:
Maybe we should keep something about `O(ln(n))` (height of the tree) vs `O(n)`.

Reply:
I thought about that and decided against it, because it seemed confusing. Is the information that the tree height is `O(ln(n))` relevant to the user of this API and why is it relevant?

Reply:
That's a really good point. Its not about a performance benefit but about the result of the operation being different. I like the binary tree construction idea, I'll add something about that to the doc. In that context the `O(ln(n))` about the height is also clear.

Reply:
Added a paragraph about building a binary search tree that includes the remark about tree height.